File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.py
import sys
input = sys.stdin.readline


K, N = map(int, input().split())
lan = []
for i in range(K):
    lan.append(int(input()))

s = 1
e = max(lan)

while s <= e:
    mid = (s+e) // 2
    l = 0
    for i in lan:
        l += i//mid
    if l >= N:
        s = mid + 1
    else:
        e = mid - 1

print(e)

File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.java
import java.util.*;
import java.io.*;

public class Main {

    static int n;
    static int k;

    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        StringTokenizer st = new StringTokenizer(br.readLine());
        k = Integer.parseInt(st.nextToken());
        n = Integer.parseInt(st.nextToken());

        long[] lan = new long[k];
        for (int i = 0; i < k; i++) {
            lan[i] = Long.parseLong(br.readLine());
        }

        long low = 1;
        long high = Arrays.stream(lan).max().getAsLong();

        while (low <= high) {
            long mid = (low + high) / 2;
            long l = 0;
            for (long i : lan)
                l += i / mid;

            if(l >= n)
                low = mid + 1;
            else
                high = mid - 1;
        }

        bw.write(String.valueOf(high));
        bw.close();
        br.close();
    }
}